File: cpp-pthread/exer07-thread-pool-main.cpp
/*
THREAD POOL & EXECUTOR SERVICE IMPLEMENTATION
*/


#include <iostream>
#include <unistd.h>
#include "exer07-thread-pool-itask.hpp"
// #include "exer07-thread-pool-v1.hpp"
// #include "exer07-thread-pool-v2a.hpp"
#include "exer07-thread-pool-v2b.hpp"



class MyTask : public ITask {
public:
    char id;

public:
    void run() override {
        std::cout << "Task " << id << " is starting" << std::endl;
        sleep(3);
        std::cout << "Task " << id << " is completed" << std::endl;
    }
};



int main() {
    constexpr int NUM_THREADS = 2;
    constexpr int NUM_TASKS = 5;


    ThreadPoolV2B threadPool;
    threadPool.init(NUM_THREADS);


    std::vector<MyTask> lstTask(NUM_TASKS);

    for (int i = 0; i < NUM_TASKS; ++i)
        lstTask[i].id = 'A' + i;


    for (auto&& task : lstTask)
        threadPool.submit(&task);

    std::cout << "All tasks are submitted" << std::endl;


    threadPool.waitTaskDone();
    std::cout << "All tasks are completed" << std::endl;


    threadPool.shutdown();


    return 0;
}

File: python/exer07_thread_pool_main.py
'''
THREAD POOL & EXECUTOR SERVICE IMPLEMENTATION
'''

import time
from exer07_thread_pool_itask import ITask
from exer07_thread_pool_v1 import ThreadPoolV1
from exer07_thread_pool_v2a import ThreadPoolV2A
from exer07_thread_pool_v2b import ThreadPoolV2B



class MyTask(ITask):
    def __init__(self, task_id: str):
        self.id = task_id

    def run(self):
        print(f'Task {self.id} is starting')
        time.sleep(3)
        print(f'Task {self.id} is completed')



NUM_THREADS = 2
NUM_TASKS = 5

thread_pool = ThreadPoolV2B(NUM_THREADS)

lsttask = [MyTask(chr(i + 65)) for i in range(NUM_TASKS)]

for task in lsttask:
    thread_pool.submit(task)

print('All tasks are submitted')

thread_pool.wait_task_done()
print('All tasks are completed')

thread_pool.shutdown()
